Add Abstractions for Mapper Output Reader and RecordReader to add Support for Integration of More Readers in the Future

aishikbh opened the pull request:
## Add Abstractions for Mapper Output Reader and RecordReaders to add Support for Integration of More Readers

This PR generalises the APIs used by the reader and record reader of Mapper Output so that adding support for additional readers and record readers for mapper output is easier in the future.

The following changes are introduced in this PR:

- Added abstraction for `GenericRowFileReader` used to read the mapper output (`MapperOutputReader`).
- Added abstraction for `GenericRowFileRecordReader` primarily used to convert back to `GenericRow` format from the mapper output  (`GenericRowMapperOutputRecordReader`).
- Included all of the APIs required by `SegmentProcessorFramework` and Reducers in the `MapperOutputReader` and `GenericRowRecordReaders` so that future Readers and RecordReaders are easily pluggable.
- Switched the implementation to the abstract ones in `SegmentProcessorFramework` and `Reducers`
-  Added a static factory for the readers to be changeable in `GenericRowFileManager` and used it to get the reader in `GenericRowFileManager`.

aishikbh committed:
Added Abstractions for Pluggability with SPF

- Added abstraction for fileReader (GenericRowReader).
- Added abstraction for GenericRow record reader (GenericRowRecordReader).
- Added support for most of the APIs required by
  SegmentProcessorFramework and Reducers except for the compare API.
- Added a static factory for the readers to be pluggable in
  GenericRowFilemanager.

aishikbh committed:
Integrate Existing Reader with SPF.

- Rename GenericRowRecordReader with GenericRowMapperOutputRecordReader
  to avoid confusion since there is already a GenericRowRecordReader.

- Use the abstractions introduced in previous commits to abstract away
  the reader specific codes including in the tests.

- Use the static factory to generate the reader in file manager.

aishikbh committed:
Rename GenericRowReader to MapperOutputReader

swaminathanmanish reviewed (commented):
If the fileReader & recordReader are specific to mapper, then consider moving the classes to mapper package, so that its not used for anything else and rename
MapperOutputReader to MapperOutputFileReader
GenericRowMapperOutputRecordReader to MapperOutputRecordReader. 
Looks good otherwise !

Could you please also link your design doc in the PR ?

aishikbh committed:
Address comments.

- Move the abstractions to mapper package since we will only use these
  specific readers and recordreaders to read mapper output.

aishikbh commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/genericrow/GenericRowFileReader.java:
```diff
@@ -85,7 +85,7 @@ public int compare(int rowId1, int rowId2) {
    * Returns a record reader for the rows within the file. Records are sorted if sort order is configured.
    */
-  public GenericRowFileRecordReader getRecordReader() {
+  public GenericRowMapperOutputRecordReader getRecordReader() {
```
This makes sense. Have moved the 3 files.

aishikbh commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/genericrow/GenericRowFileManager.java:
```diff
@@ -99,12 +101,18 @@ public void closeFileWriter()
       Preconditions.checkState(_offsetFile.exists(), "Record offset file: %s does not exist", _offsetFile);
       Preconditions.checkState(_dataFile.exists(), "Record data file: %s does not exist", _dataFile);
-      _fileReader = new GenericRowFileReader(_offsetFile, _dataFile, _fieldSpecs, _includeNullFields, _numSortFields);
+      Map<String, Object> params = new HashMap<>();
```
Unfortunately this can vary across readers, that's why decided to put it in this way
